src/jit/codegenxarch.cpp

```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+                    // For this case, we must adjust the stack and generate stack-relative stores rather than pushes.
+                    // Adjust the stack pointer to the next slot boundary.
+                    inst_RV_IV(INS_sub, REG_SPBASE, adjustment, EA_PTRSIZE);
+                    currentOffset -= adjustment;
+                    genStackLevel += adjustment;
+                }
             }
 
-            // We can use a push instruction for any slot-like field.
-            //
-            // NOTE: if the field is of GC type, we must use a push instruction, since the emitter is not otherwise
-            // able to detect stores into the outgoing argument area of the stack on x86.
-            const bool usePush = !preAdjustedStack && fieldIsSlot;
-            assert(usePush || !varTypeIsGC(fieldType));
-
-            // Adjust the stack if necessary. If we are going to generate a `push` instruction, this moves the stack
-            // pointer to (fieldOffset + sizeof(fieldType)) to account for the `push`.
-            const int fieldSize     = genTypeSize(fieldType);
-            const int desiredOffset = current->gtFieldOffset + (usePush ? fieldSize : 0);
-            if (currentOffset > desiredOffset)
-            {
-                assert(!preAdjustedStack);
-
-                // The GC encoder requires that the stack remain 4-byte aligned at all times. Round the adjustment up
-                // to the next multiple of 4. If we are going to generate a `push` instruction, the adjustment must
-                // not require rounding.
-                const int adjustment = roundUp(currentOffset - desiredOffset, 4);
-                assert(!usePush || (adjustment == (currentOffset - desiredOffset)));
-                inst_RV_IV(INS_sub, REG_SPBASE, adjustment, EA_PTRSIZE);
-                currentOffset -= adjustment;
-                genStackLevel += adjustment;
-            }
-
-            // Note that the argReg may not be the lcl->gtRegNum, if it has been copied
-            // or reloaded to a different register.
-            const regNumber argReg = fieldNode->gtRegNum;
-            if (usePush)
-            {
-                // Adjust the stack if necessary and push the field.
-                // Push the field.
-                inst_RV(INS_push, argReg, fieldType, emitTypeSize(fieldType));
-                currentOffset -= fieldSize;
-                genStackLevel += fieldSize;
+            // Consume the register, if any, for this field. Note that genConsumeRegs() will appropriately
+            // update the liveness info for a lclVar that has been marked RegOptional, which hasn't been
+            // assigned a register, and which is therefore contained.
+            // Unlike genConsumeReg(), it handles the case where no registers are being consumed.
+            genConsumeRegs(fieldNode);
+            regNumber argReg = fieldNode->gtRegNum;
```

Comment:
Also, is there a reason to do this here rather than prior to the stack adjustment? If we did this earlier, we could fold the `if (!m_pushStkArg)` code below into the `else` block of the `if (fieldIsSlot)` statement above.